src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -236,64 +237,170 @@ object DeployPlugin extends AutoPlugin {
-        } else {
-          Seq()
-        }
-
-        val restartCommand = Seq.concat(
-          sshCommand,
-          Seq(deployHost, restartScript, "restart"),
-          restartArgs
-        )
+        Seq("ssh", "-i", keyFile, "-l", hostConfig.sshUser)
+      }
 
-        // Run commands within a Try so we can collect errors using Future.fold.
-        // TODO(danm): Use ProcessIO to capture and report the actual stderr of failures here.
-        Try {
-          log.info("Running " + quotedRsync + " . . .")
-          if (Process(rsyncCommand).! != 0) {
-            sys.error(s"Error running rsync to host '$deployHost'.")
-          }
+      val replicas = deployConfig.replicaOverridesByHost(hostConfig)
+      val numReplicas = replicas.length
+
+      // First, stop and remove any replicas on the remote that won't be restarted
+      // as part of the deploy.
+      val (deployParent, deployTarget) = baseDeployDir.splitAt(baseDeployDir.lastIndexOf('/'))
+      // Strip leading / from deployTarget so we can match names on it.
+      val namePattern = deployTarget.tail
+      // Build a regex matching the replicas we will eventually update / restart as part of this
+      // process, and therefore DON'T want to stop and remove.
+      val restartedPattern = if (numReplicas == 1) {
+        namePattern
+      } else {
+        s"$namePattern-[1-$numReplicas]"
+      }
+      // Use find -exec to locate, stop, and remove stale replicas that won't be updated as part
+      // of the current deploy.
+      val findCommand = Seq(
+        "find",
+        s"$deployParent/*",
+        "-prune",
+        "-type", "d",
+        "-name", "\"" + s"$namePattern*" + "\"",
+        "-not", "-name", "\"" + restartedPattern + "\"",
+        "-exec", s"{}/bin/$namePattern.sh stop \\;",
+        "-exec", "rm -r {} \\;"
+      )
+      val stopCommand = Seq.concat(sshCommand, Seq(deployHost), findCommand)
+      // Shell-friendly version of find command, for logging.
+      val quotedStopCommand = Seq.concat(
+        sshCommand,
+        Seq(deployHost, findCommand.mkString("'", " ", "'"))
+      )
+
+      log.info("Running " + quotedStopCommand.mkString(" ") + " . . .")
+      if (Process(stopCommand).! != 0) {
+        sys.error(s"Error while trying to tear down stale replicas on '$deployHost'.")
+      }
 
-          log.info("Running " + restartCommand.mkString(" ") + " . . .")
-          if (Process(restartCommand).! != 0) {
-            sys.error(s"Error running restart command on host '$deployHost'.")
+      replicas.zipWithIndex map {
+        case (replicaConfig, i) => {
+          Future {
+            // Determine the specific remote directory for this replica.
+            val deployDirectory = if (numReplicas > 1) {
+              s"$baseDeployDir-${i + 1}"
+            } else {
+              baseDeployDir
+            }
+
+            // Command to pass to rsync's "rsync-path" flag to ensure creation of the remote dir.
```

Comment:
The `--rsync-path` is just a <del>trick</del> hack to make the `mkdir` and `rsync` run in a single remote call.

I do seem to have a tendency to over-optimize.